pinot-core/src/main/java/com/linkedin/pinot/core/common/DataFetcher.java

```diff
@@ -33,47 +35,49 @@
 public class DataFetcher {
   private static final BlockId BLOCK_ZERO = new BlockId(0);
 
-  private final Map<String, BaseOperator> _columnToDataSourceMap;
   private final Map<String, Dictionary> _columnToDictionaryMap;
   private final Map<String, BlockValSet> _columnToBlockValSetMap;
   private final Map<String, BlockValIterator> _columnToBlockValIteratorMap;
   private final Map<String, BlockMetadata> _columnToBlockMetadataMap;
 
+  // Map from MV column name to max number of entries for the column.
+  private final Map<String, Integer> _columnToMaxNumMultiValuesMap;
+
+  // Re-usable array for all dictionary ids in the block, of a single valued column
+  private final Map<String, int[]> _columnToDictIdsMap;
```

Comment:
Data fetcher is not a cache, and should not maintain any reusable array or fetched data inside. This _columnToDictIdsMap is a block level dict ids cache, which definitely should not inside data fetcher. Data fetcher should be on top of the whole segment, but not a block. For block level data / cache, we should put them into BlockCache.